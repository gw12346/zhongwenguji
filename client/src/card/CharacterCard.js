import React from 'react';
import cx from 'classnames';
import formatPinyin from '../util/formatPinyin';
import style from './CharacterCard.module.scss'

export default (props) => {
  const {
    showDefinition,
    character,
    pinyin,
    definition,
    score,
    className
  } = props;

  const finalClassName = cx(style.container, {
    [style.showDefinition]: showDefinition,
    [style.markedUnknown]: score === 0,
    [style.markedKnown]: score === 1
  }, className);

  return (
    <span className={finalClassName}>
      {showDefinition ? (
        <>
          <span className={style.pinyin}>
            {formatPinyin(pinyin)}
          </span>
          <span className={style.definition}>
            {(definition || '(No definition available)').replace(/\//g, ' / ')}
          </span>
        </>
      ) : (
        <span className={style.character}>
          {character}
        </span>
      )}
    </span>
  )
};
