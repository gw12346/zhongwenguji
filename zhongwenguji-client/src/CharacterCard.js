import React from 'react';
import { formatPinyin } from './util';

export default ({ showDefinition = false, data: { i, c, p, d } }) => (
  <span className="TestCard">
    {showDefinition ? (
      <>
        <span className="TestCard__pinyin">{formatPinyin(p)}</span>
        <span className="TestCard__definition">{d.replace(/\//g, ' / ')}</span>
      </>
    ) : (
      <span className="TestCard__character">{c}</span>
    )}
  </span>
);
