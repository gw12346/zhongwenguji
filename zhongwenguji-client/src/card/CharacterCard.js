import React from 'react';
import cx from 'classnames';
import formatPinyin from '../util/formatPinyin';
import style from './CharacterCard.module.scss'

export default (props) => {
  const {
    showDefinition,
    character,
    pinyin,
    definition,
    score
  } = props;

  const className = cx(style.container, {
    [style.showDefinition]: showDefinition,
    [style.markedUnknown]: score === 0,
    [style.markedKnown]: score === 1
  });

  return (
    <span className={className}>
      {showDefinition ? (
        <>
          <span className={style.pinyin}>
            {formatPinyin(pinyin)}
          </span>
          <span className={style.definition}>
            {(definition || '(No definition available)').replace(/\//g, ' / ')}
          </span>
        </>
      ) : (
        <span className={style.character}>
          {character}
        </span>
      )}
    </span>
  )
};
