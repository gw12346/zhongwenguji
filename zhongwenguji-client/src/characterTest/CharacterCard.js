import React from 'react';
import formatPinyin from '../util/formatPinyin';

export default ({ showDefinition = false, card: { i, c, p, d } }) => (
  <span className={['TestCard', showDefinition ? 'TestCard--showDefinition' : ''].join(' ')}>
    {showDefinition ? (
      <>
        <span className="TestCard__pinyin">{formatPinyin(p)}</span>
        <span className="TestCard__definition">{d.replace(/\//g, ' / ')}</span>
      </>
    ) : (
      <span className="TestCard__character">{c}</span>
    )}
  </span>
);
