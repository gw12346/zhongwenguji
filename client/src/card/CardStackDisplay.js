import React, { Component } from 'react';
import cx from 'classnames';
import { CSSTransitionGroup } from 'react-transition-group';
import CharacterCard from './CharacterCard';
import style from './CardStackDisplay.module.scss';
import PropTypes from "prop-types";
import noop from "../util/noop";

function isAncestor(ancestor, child) {
  let check = child;
  while (check) {
    if (check === ancestor) {
      return true;
    }
    check = check.parentElement;
  }
  return false;
}

export default class CardStackDisplay extends Component {
  static propTypes = {
    onDiscardLeft: PropTypes.func,
    onDiscardRight: PropTypes.func,
  };

  static defaultProps = {
    onDiscardLeft: noop,
    onDiscardRight: noop
  };

  constructor() {
    super();
    this.state = { offsetVector: [0, 0, 0] };
    this.touchAreaRef = React.createRef();
  }

  getActiveTouch = (event) => {
    return this.state.activeTouch && [].find.call(event.changedTouches, touch => touch.identifier === this.state.activeTouch.identifier);
  };

  setOffsetVector(offsetVector, disableTransition = false) {
    this.setState({ offsetVector, disableTransition });
    if (disableTransition) {
      setTimeout(() => this.setState({ disableTransition: false }), 50);
    }
  }

  handleTouchStart = (event) => {
    const isTouchArea = isAncestor(this.touchAreaRef.current, event.target);
    const isNewFirstTouch = event.touches.length === 1;
    const isTouchActive = !!this.state.activeTouch;
    if (isTouchArea && (!isTouchActive || (isTouchActive && isNewFirstTouch))) {
      event.preventDefault();
      const { identifier, clientX, clientY } = event.changedTouches[0];
      this.setState({ activeTouch: { identifier, clientX, clientY } });
    }
  };

  handleTouchMove = (event) => {
    const activeTouch = this.getActiveTouch(event);
    if (activeTouch) {
      const dx = activeTouch.clientX - this.state.activeTouch.clientX;
      const dy = Math.abs(dx / 10);
      const dth = Math.sin(Math.PI * dx / 2000) * 180 / Math.PI;
      this.setOffsetVector([dx, dy, dth]);
    }
  };

  handleTouchEnd = (event) => {
    if (this.getActiveTouch(event)) {
      const { discardThreshold } = this.props;
      const [ dx ] = this.state.offsetVector;
      if (dx > discardThreshold) {
        this.props.onDiscardRight();
      } else if (-dx > discardThreshold) {
        this.props.onDiscardLeft();
      } else {
        this.setOffsetVector([0, 0, 0]);
      }
      this.setState({ activeTouch: null });
    }
  };

  componentDidMount() {
    document.addEventListener('touchstart', this.handleTouchStart, { passive: false });
    document.addEventListener('touchend', this.handleTouchEnd);
    document.addEventListener('touchcancel', this.handleTouchEnd);
    document.addEventListener('touchmove', this.handleTouchMove);
  }

  componentWillUpdate(newProps) {
    if (this.props.card
      && newProps.card
      && this.props.card.index !== newProps.card.index
    ) {
      this.setOffsetVector([0, 0, 0]);
    }
  }

  componentWillUnmount() {
    document.removeEventListener('touchstart', this.handleTouchStart);
    document.removeEventListener('touchend', this.handleTouchEnd);
    document.removeEventListener('touchcancel', this.handleTouchEnd);
    document.removeEventListener('touchmove', this.handleTouchMove);
  }

  getPositionOffsetStyles() {
    return {
      left: `${Math.round(this.state.offsetVector[0])}px`,
      top: `${Math.round(this.state.offsetVector[1])}px`,
      transform: `rotate(${this.state.offsetVector[2]}deg)`
    };
  }

  renderX() {
    const { card, showDefinition, discardThreshold } = this.props;
    const { disableTransition } = this.state;
    const touchActive = !!this.state.activeTouch;
    const [ dx ] = this.state.offsetVector;

    return (
      <div
        className={cx(style.touchArea, { [style.touchAreaSnap]: disableTransition || touchActive })}
        style={this.getPositionOffsetStyles()}
      >
        {card.index >= 0 ? (
          <CharacterCard
            {...card}
            className={cx(style.card, {
              [style.predictDiscardRight]: dx > discardThreshold,
              [style.predictDiscardLeft]: dx < -discardThreshold
            })}
            showDefinition={showDefinition}
          />
        ) : null}
      </div>
    );
  }

  render() {
    const { card } = this.props;
    return (
      <div className={style.cardContainer}>
        <div className={style.stackBase} ref={this.touchAreaRef}>
          <CSSTransitionGroup
            component="div"
            transitionName={style}
            transitionEnterTimeout={300}
            transitionLeaveTimeout={300}
          >
            {card ? (
              <div
                key={card.index}
                className={cx(style.animationContainer, {
                  [style.discardRight]: card.score === 1,
                  [style.discardLeft]: card.score === 0
                })}>
                {this.renderX()}
              </div>
            ) : null}
          </CSSTransitionGroup>
        </div>
      </div>
    );
  }
}
