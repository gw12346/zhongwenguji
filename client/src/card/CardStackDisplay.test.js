import React from 'react';
import Renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { JSDOM } from 'jsdom';
import CharacterCard from './CharacterCard';
import CardStackDisplay from './CardStackDisplay';

const classNames = (element) => {
  if (!element || !element.props) {
    throw new Error('A valid element was not provided');
  }
  return (element.props.className || '').split(/\s+/);
};
const hasClassName = className => (element) => classNames(element).indexOf(className) >= 0;

const BASE_CARD = { index: 1, score: NaN };
const DISCARD_THRESHOLD = 50;

describe('CardStackDisplay', () => {
  let eventListener;
  let component;
  let animationContainer;
  let touchArea;
  let characterCard;

  function setupEventListener() {
    eventListener = {};
    document.addEventListener = jest.fn((eventType, callback) => {
      eventListener[eventType] = callback;
    });
  }

  function setup(props) {
    setupEventListener();
    component = Renderer.create(
      <CardStackDisplay
        card={BASE_CARD}
        discardThreshold={DISCARD_THRESHOLD}
        {...props}
      />
    );
    animationContainer = component.root.find(hasClassName('animationContainer'));
    touchArea = component.root.find(hasClassName('touchArea'));
    characterCard = component.root.findByType(CharacterCard);
  }

  function setupEnzyme(props) {
    setupEventListener();
    component = mount(
      <CardStackDisplay
        card={BASE_CARD}
        discardThreshold={DISCARD_THRESHOLD}
        {...props}
      />
    );
    animationContainer = component.find('.animationContainer');
    touchArea = component.find('.touchArea');
    characterCard = component.find(CharacterCard);
  }

  it('renders', () => {
    setup();
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('animation', () => {
    it('applies the discardLeft style', () => {
      setup({ card: { ...BASE_CARD, score: 0 } });
      expect(classNames(animationContainer))
        .toEqual(expect.arrayContaining(['discardLeft']));
    });

    it('applies the discardRight style', () => {
      setup({ card: { ...BASE_CARD, score: 1 } });
      expect(classNames(animationContainer))
        .toEqual(expect.arrayContaining(['discardRight']));
    });
  });

  describe('swiping left', () => {
    const applyTouch = ({ x: clientX, y: clientY, identifier = 0 }) => ({
      touches: [{ identifier, clientX, clientY }],
      changedTouches: [{ identifier, clientX, clientY }]
    });

    beforeAll(() => {
      setupEnzyme();
    });

    it('starts', () => {
      const preventDefault = jest.fn();
      eventListener['touchstart']({
        target: touchArea.getDOMNode(),
        preventDefault,
        ...applyTouch({ x: 150, y: 150 })
      });
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it('moves', () => {
      eventListener['touchmove']({
        target: touchArea.getDOMNode(),
        ...applyTouch({ x: 99, y: 150 })
      });

      // TODO: write assertion
    });
  });
});
