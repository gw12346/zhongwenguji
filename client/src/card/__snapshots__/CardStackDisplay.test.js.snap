// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardStackDisplay renders 1`] = `
<div
  className="cardContainer"
>
  <div
    className="stackBase"
  >
    <div>
      <div
        className="animationContainer"
      >
        <div
          className="touchArea"
          style={
            Object {
              "left": "0px",
              "top": "0px",
              "transform": "rotate(0deg)",
            }
          }
        >
          <span
            className="container card"
          >
            <span
              className="character"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
